staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 			continue
 		}
 		if evalResult != types.True {
-			if len(rule.Message) != 0 {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, rule.Message))
+			if compiled.MessageExpression != nil {
+				messageExpression, newRemainingBudget, msgErr := evalMessageExpression(ctx, compiled.MessageExpression, rule.MessageExpression, activation, remainingBudget)
```

Comment:
This is tricky here. `runtimeCostErr` you defined actually includes two types of runtimeCostErr: 
- cost failed to retrieve. 
- cost exceeds budge. 

However, the `evalErr` you defined also includes two types of err: 
- cost per expression failed. 
- eval failed.  

In current code, both `cost per expression failed.` and `eval failed` will lead to warning and using message/rule. I guess we would like to keep the behavior same as existing one which is `cost per expression failed` will stop running following expression and return err.

Reply:
Right, the problem is using `continue` and not an immediate `return`. I'll move it to be in line with current behavior and add some tests that check for runtime cost failures.

Reply:
(marking as resolved - cost per expression and eval failure should be properly tested now)